lib/streamlit/elements/widgets/text_widgets.py

```diff
@@ -48,45 +50,85 @@
 
 @dataclass
 class TextInputSerde:
-    value: SupportsStr
+    value: str | None
 
-    def deserialize(self, ui_value: Optional[str], widget_id: str = "") -> str:
-        return str(ui_value if ui_value is not None else self.value)
+    def deserialize(self, ui_value: str | None, widget_id: str = "") -> str | None:
+        return ui_value if ui_value is not None else self.value
 
-    def serialize(self, v: str) -> str:
+    def serialize(self, v: str | None) -> str | None:
         return v
 
 
 @dataclass
 class TextAreaSerde:
-    value: SupportsStr
+    value: str | None
 
-    def deserialize(self, ui_value: Optional[str], widget_id: str = "") -> str:
-        return str(ui_value if ui_value is not None else self.value)
+    def deserialize(self, ui_value: str | None, widget_id: str = "") -> str | None:
+        return ui_value if ui_value is not None else self.value
 
-    def serialize(self, v: str) -> str:
+    def serialize(self, v: str | None) -> str | None:
         return v
 
 
 class TextWidgetsMixin:
-    @gather_metrics("text_input")
+    @overload
     def text_input(
         self,
         label: str,
-        value: SupportsStr = "",
-        max_chars: Optional[int] = None,
-        key: Optional[Key] = None,
+        value: str = "",
+        max_chars: int | None = None,
+        key: Key | None = None,
         type: Literal["default", "password"] = "default",
-        help: Optional[str] = None,
-        autocomplete: Optional[str] = None,
-        on_change: Optional[WidgetCallback] = None,
-        args: Optional[WidgetArgs] = None,
-        kwargs: Optional[WidgetKwargs] = None,
+        help: str | None = None,
+        autocomplete: str | None = None,
+        on_change: WidgetCallback | None = None,
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
         *,  # keyword-only arguments:
-        placeholder: Optional[str] = None,
+        placeholder: str | None = None,
         disabled: bool = False,
         label_visibility: LabelVisibility = "visible",
     ) -> str:
+        pass
+
+    @overload
+    def text_input(
```

Comment:
Hm, would it make more sense for the first `@overload` for this widget to take a value with type `SupportsStr` and return a `str`, then the second `@overload` is where value has type `None` and returns `None` (it might be necessary to make the `None` case the first overload in the list in case order is important)

I think this should be equivalent but a bit more clear

(same comment for `text_area` below)

Reply:
That was my initial try as well, but it's giving me a mypy error:

```
Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader  [misc]
```

and if I change the order:

```
Overloaded function signatures 1 and 2 overlap with incompatible return types
```

The main issue seems to be with the `SupportsStr` protocol since this also includes `None` and it requires non-overlapping types here :( With the workaround by adding `str` explicitly, I can at least provide optimized typing for all uses with a normal string (which are most uses anyways).

Reply:
I played around today a bit more with the overloads but didn't really find a way to get it to work other than the current workaround :(